src/linear_algebra/orientation.lean

```diff
@@ -0,0 +1,384 @@
+the positive orientation.
+
+## Implementation notes
+
+`orientation` is defined for an arbitrary index type, but the main intended use case is when
+that index type is a `fintype` and there exists a basis of the same cardinality.
+
+## References
+
+* https://en.wikipedia.org/wiki/Orientation_(vector_space)
+
+-/
+
+noncomputable theory
+
+section ordered_comm_semiring
+
+variables (R : Type*) [ordered_comm_semiring R]
+variables {M : Type*} [add_comm_monoid M] [module R M]
+variables (ι : Type*) [decidable_eq ι]
+
+/-- Two vectors are in the same ray if some positive multiples of them are equal (in the typical
+case over a field, this means each is a positive multiple of the other).  Over a field, this
+is equivalent to `mul_action.orbit_rel`. -/
+def same_ray (v₁ v₂ : M) : Prop :=
+∃ (r₁ r₂ : R), 0 < r₁ ∧ 0 < r₂ ∧ r₁ • v₁ = r₂ • v₂
+
+variables (M)
+
+/-- Nonzero vectors, as used to define rays. -/
+@[reducible] def ray_vector := {v : M // v ≠ 0}
+
+/-- `same_ray` is symmetric. -/
+lemma symmetric_same_ray : symmetric (same_ray R : M → M → Prop) :=
+λ _ _ ⟨r₁, r₂, hr₁, hr₂, h⟩, ⟨r₂, r₁, hr₂, hr₁, h.symm⟩
+
+/-- `same_ray` is transitive. -/
+lemma transitive_same_ray :
+  transitive (same_ray R : M → M → Prop) :=
+λ _ _ _ ⟨r₁, r₂, hr₁, hr₂, h₁⟩ ⟨r₃, r₄, hr₃, hr₄, h₂⟩,
+  ⟨r₃ * r₁, r₂ * r₄, mul_pos hr₃ hr₁, mul_pos hr₂ hr₄,
+   by rw [mul_smul, mul_smul, h₁, ←h₂, smul_comm]⟩
+
+variables [nontrivial R]
+
+/-- `same_ray` is reflexive. -/
+lemma reflexive_same_ray :
+  reflexive (same_ray R : M → M → Prop) :=
+λ _, ⟨1, 1, zero_lt_one, zero_lt_one, rfl⟩
+
+/-- `same_ray` is an equivalence relation. -/
+lemma equivalence_same_ray :
+  equivalence (same_ray R : M → M → Prop) :=
+⟨reflexive_same_ray R M, symmetric_same_ray R M, transitive_same_ray R M⟩
+
+variables {R M}
+
+/-- A vector is in the same ray as a positive multiple of itself. -/
+lemma same_ray_pos_smul_right (v : M) {r : R} (h : 0 < r) :
+  same_ray R v (r • v) :=
+⟨r, 1, h, zero_lt_one, (one_smul _ _).symm⟩
+
+/-- A vector is in the same ray as a positive multiple of one it is in the same ray as. -/
+lemma same_ray.pos_smul_right {v₁ v₂ : M} {r : R} (h : same_ray R v₁ v₂) (hr : 0 < r) :
+  same_ray R v₁ (r • v₂) :=
+transitive_same_ray R M h (same_ray_pos_smul_right v₂ hr)
+
+/-- A positive multiple of a vector is in the same ray as that vector. -/
+lemma same_ray_pos_smul_left (v : M) {r : R} (h : 0 < r) :
+  same_ray R (r • v) v :=
+⟨1, r, zero_lt_one, h, one_smul _ _⟩
+
+/-- A positive multiple of a vector is in the same ray as one it is in the same ray as. -/
+lemma same_ray.pos_smul_left {v₁ v₂ : M} {r : R} (h : same_ray R v₁ v₂) (hr : 0 < r) :
+  same_ray R (r • v₁) v₂ :=
+transitive_same_ray R M (same_ray_pos_smul_left v₁ hr) h
+
+variables (R M)
+
+/-- The setoid of the `same_ray` relation for the subtype of nonzero vectors. -/
+def same_ray_setoid : setoid (ray_vector M) :=
+{ r := λ v₁ v₂, same_ray R (v₁ : M) v₂,
+  iseqv := equivalence.comap (equivalence_same_ray R M) _ }
```

Comment:
I'd be inclined to add both setoids here:
```suggestion
/-- The setoid of the `same_ray` relation for elements of a module. -/
def same_ray_setoid [nontrivial R] : setoid M :=
{ r := λ v₁ v₂, same_ray R v₁ v₂, iseqv := equivalence_same_ray R M }

/-- The setoid of the `same_ray` relation for the subtype of nonzero vectors. -/
def ray_vector.same_ray [nontrivial R] : setoid (ray_vector M) :=
(same_ray_setoid R M).comap coe
```
I think it then makes sense to move all the `ray_vector` stuff down below the `same_ray` stuff

Reply:
I've moved the `ray_vector` definition, but I'm not clear what else you might be suggesting moving (the split by type class requirements means `same_ray` lemmas for `ordered_comm_ring` and `linear_ordered_comm_ring` are below `ray_vector` lemmas for `ordered_comm_semiring`, unless the sections for each type class requirement get split up themselves).